---
title: "fecal-mifish-decontamination"
author: "diana baetscher"
date: "2024-10-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Borrowing from Kim's Bering Sea data decon procedure...


```{r load-libraries}
library(tidyverse)

collapsed_taxonomy <- read_csv("csv_outputs/inrange_collapsed_fecal_mifish_taxonomy.csv")

collapsed_taxonomy %>%
  filter(str_detect(taxon, "Oncorhynchus"))

```
Those observations are in the dataframe, but I'm adding a different dataframe!



```{r load-sample-info}
sample_metadata <- readxl::read_xlsx("../data/NFS_metadata_20241001.xlsx")

#illumina output changed "_" to "-"
# sample_metadata$sample_ID <- gsub("_", "-", sample_metadata$sample_ID) 
# sample_metadata$sample_ID_date <- gsub("_", "-", sample_metadata$sample_ID_date) 
```

check sequence table outputs
```{r}
asv_table <- read.csv("csv_outputs/NFSfecal_MiFish_ASVtable.csv") %>%
  rename(sample = X)

# reformat
sample_df <- asv_table %>%
  pivot_longer(2:length(asv_table), names_to = "ASV", values_to = "reads") %>%
  filter(reads > 0) %>%
  separate(sample, into = c("ablg", "plate", "replicate"), remove = F)

head(sample_df)
```


# account for likely contaminants 

- tag-jumping
- negative PCR controls
- field negatives


## Step 1. Account for tag-jumping by using the positive controls 

subtract the proportion of reads that jumped into the positive control samples from each environmental sample 

```{r tag-jumping-by-ASVs-in-PC}
# on a per-plate basis, what % of each ASV should be subtracted from the samples?
asv_prop_df <- sample_df %>%
  group_by(sample) %>%
  mutate(total_reads = sum(reads)) %>%
  mutate(asv_prop = reads/total_reads)
  
asv_prop_df %>%
  filter(ablg == "PC") %>%
  ggplot(aes(x = sample, y = asv_prop, fill = ASV)) +
  geom_bar(stat = "identity") +
  facet_grid(row = vars(replicate)) +
  theme_minimal()

```
```{r which-asv-proportions-to-remove}
# based on asvs in the PC
asv_prop_to_remove <- asv_prop_df %>%
  filter(ablg == "PC") %>%
  ungroup() %>%
  select(plate, ASV, asv_prop) %>%
  rename(prop_to_subtract = asv_prop)

```


```{r remove-asvs-props-from-PC-in-samples}
pc_filtered_sample_df <- asv_prop_df %>%
  left_join(., asv_prop_to_remove) %>%
  mutate(new_asv_prop = ifelse(!is.na(prop_to_subtract), asv_prop-prop_to_subtract, asv_prop)) %>%
  mutate(new_asv_prop = ifelse(new_asv_prop < 0, 0, new_asv_prop)) %>% # remove negative numbers
  #filter(new_asv_prop > 0) %>%
  mutate(reads_removed = prop_to_subtract*total_reads) %>%
  select(-prop_to_subtract, -asv_prop)

pc_filtered_sample_df
```

Reads removed by tag-jumping:
```{r summarize-reads-removed}
pc_filtered_sample_df %>%
  filter(!is.na(reads_removed))
```





this is a summary of the number of reads removed by ASV and sample_ID
```{r summarize}
decontaminated_1 <- pc_filtered_sample_df %>%
  dplyr::select(sample, ASV, reads_removed) %>%
  filter(!is.na(replicate) &
           !is.na(reads_removed)) %>%
  group_by(ASV) %>%
  summarise(mean_reads = mean(reads_removed),
            reads_q.05 = quantile(reads_removed, probs=0.05),
            median_q.5 = median(reads_removed),
            reads_q.95 = quantile(reads_removed, probs=0.95)) %>%
  filter(mean_reads > 0) %>%
  filter(ASV != "ASV16") # remove sturgeon

decontaminated_1  

```


## Step 2. Account for contaminants in positive and negative controls 

next we will remove ASVs that only occur in controls and not in environmental samples. 

number of reads
```{r}
reads_per_type_ASV <- pc_filtered_sample_df %>%
  left_join(., sample_metadata, by = c("ablg" = "extraction_ID")) %>% 
  mutate(sample_type = ifelse(is.na(sample_type) & ablg == "NC", "pcr_blank", sample_type)) %>%
  mutate(sample_type = ifelse(is.na(sample_type) & ablg == "PC", "positive_control", sample_type)) %>%
  group_by(ASV, sample_type) %>%
  summarize(TotalReadsPerASV = sum(reads, na.rm = TRUE)) %>%
  arrange(ASV)

```

what ASVs have no reads in samples, but reads in the controls? 
```{r}
not_in_samples <- reads_per_type_ASV %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerASV")) %>%
  filter(is.na(sample)) %>% # here, Kim had a numeric, but mine are NA
  replace(is.na(.), 0)

not_in_samples
```


what ASVs do have reads in samples, but more reads in the controls? 
```{r}
more_in_pcr_blanks <- reads_per_type_ASV %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerASV")) %>%
  filter(sample > 1) %>%
  filter(pcr_blank > sample)
head(more_in_pcr_blanks)

more_in_extraction_blanks <- reads_per_type_ASV %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerASV")) %>%
  filter(sample > 1) %>%
  filter(extraction_blank > sample)
head(more_in_extraction_blanks)

more_in_pc_blanks <- reads_per_type_ASV %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerASV")) %>%
  filter(sample > 1) %>%
  filter(positive_control > sample)
head(more_in_pc_blanks)

more_in_fb_blanks <- reads_per_type_ASV %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerASV")) %>%
  filter(sample > 1) %>%
  filter(field_blank > sample)
head(more_in_fb_blanks)
```

remove these from the asv table
```{r}
asv_table_filter2 <- pc_filtered_sample_df %>%
  filter(!ASV %in% not_in_samples$ASV) %>%
  select(-reads_removed, -total_reads, -new_asv_prop)
 # filter(!ASV %in% more_in_pcr_blanks$ASV) %>%
  #filter(!ASV %in% more_in_extraction_blanks$ASV) %>%
  #filter(!ASV %in% more_in_fb_blanks$ASV)

# Let's stick with ASVs that are just in the controls for filtering
```


## Step 3. Remove ASVs without taxonomic ID 

now lets see how many of these ASVs have taxonomic IDs (these are not final tax ids)
```{r output-from-blastn-and-taxonkit}
taxonomy <- read_csv("csv_outputs/inrange_collapsed_fecal_mifish_taxonomy.csv")

combined_df <- asv_table_filter2 %>%
  left_join(., taxonomy) 
```



```{r merge-data-frames-remove-NAs}
combined_df %>%
  filter(is.na(taxon)) %>%
  ungroup() %>%
  select(ASV) %>%
  unique()

```
188 ASVs without taxonomy (removed).

Some of those are selectively removed because they're off-target (i.e., primates, bacteria, etc.)

```{r remove-ASVs-without-taxonomy}
combined_df_noNA <- combined_df %>%
  filter(!is.na(taxon)) 

combined_df_noNA %>%
  group_by(taxon) %>%
  summarise(sum(reads))

```


```{r filter-controls-analyze-samples}

clean_df_filtered <- combined_df_noNA %>%
  left_join(., sample_metadata, by = c("ablg" = "extraction_ID")) %>%
  filter(sample_type == "sample") %>%
  filter(!taxon %in% c("Acipenser fulvescens", "Acipenser", "Callorhinus ursinus", "Eumetopias jubatus")) %>%
  group_by(sample, taxon) %>%
  mutate(taxon_reads = sum(reads)) %>%
  select(-ASV, -reads) %>%
  unique() %>%
  ungroup() %>%
  group_by(sample) %>%
  mutate(total_reads = sum(taxon_reads)) %>%
  select(sample, ablg, plate, replicate, taxon_reads, taxon, total_reads, everything()) %>%
  mutate(prop_taxon = taxon_reads/total_reads)
  

clean_df_filtered %>%
  ggplot(aes(x = ablg, y = prop_taxon, fill = taxon)) +
  geom_bar(stat = "identity") +
  facet_grid(cols = vars(replicate), rows = vars(plate)) +
  theme_minimal() +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 90)
  ) 

```



### read in data from hard parts
(from Katie Luxa and Tonya Zeppelin at MML)


```{r hard-parts-data}
hardparts <- readxl::read_xlsx("../data/2023 NFS Bogoslof hard parts.xlsx")

# just keep the fishes 
# i.e., remove squids
hardparts_df <- hardparts %>%
  # select(ScientificName) %>%
  # unique() %>%
  filter(!ScientificName %in% c("Gonatidae", "Gonatopsis borealis/Berryteuthis magister", "Gonatus sp.", "NO PREY REMAINS", "Decapodiformes")) %>%
  rename(hardparts = ScientificName) %>%
  mutate(hardparts = ifelse(hardparts == "Northern lampfish", "Stenobrachius leucopsaurus", hardparts))

```
9 species of fishes.


```{r}
clean_df_filtered %>%
  #left_join(., hardparts_df, by = c("alternative_ID" = "DNASample")) %>%
  anti_join(., hardparts_df, by = c("alternative_ID" = "DNASample")) %>%
  filter(source != "fecal loop") %>%
  select(alternative_ID) %>%
  unique()
  
  
  # filter(!is.na(hardparts)) %>%
  # ggplot(aes(x = alternative_ID, y = prop_taxon, fill = taxon)) +
  # geom_bar(stat = "identity") +
  # facet_grid(cols = vars(replicate), rows = vars(plate)) +
  # theme_minimal() +
  # theme(
  #   legend.position = "none",
  #   axis.text.x = element_text(angle = 90)
  # ) 

```
From the hard parts, apparently we're just getting detections (y/n) for each of the species?

```{r}
hardparts_df %>%
  ggplot(aes(x = hardparts, fill = hardparts)) +
  geom_bar(stat = "count") +
  #facet_grid(cols = vars(replicate), rows = vars(plate)) +
  theme_minimal() +
  theme(
    #legend.position = "none",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  ) 
  
```
Could we do presence/absence linear regression, hardparts vs. metabarcoding by species detections in the dataset?


Apply Bray-Curtis to metabarcoding replicates, then, assuming everything is similar, use the P/A data for comparison with hardparts.
```{r}
library(vegan)

# change df format for Bray-Curtis analysis
clean_df_filtered

```

```{r}
clean_df_filtered %>%
  arrange(taxon_reads) %>%
  # retaining taxa with 1 read seems crazy.
  ggplot(aes(taxon_reads)) +
  geom_histogram()
```
```{r}
min_read_df <- clean_df_filtered %>%
  filter(taxon_reads > 1)
```




```{r}
## simple bray-curtis (downsampling, VRP)
sample.df <- min_read_df %>%
    #filter(ablg == "e03719") %>% # I think I need to do this sample-by-sample?
    select(sample, taxon, taxon_reads) %>%
    unique() %>% # ensure there are no duplicates because they will cause the pivot to fail later
    group_by(taxon, sample) %>%
    summarise(count = sum(taxon_reads))
  
  ### community-by-species matrix
  # To run the Bray-Curtis and NMDS, we will use the function metaMDS. The function requires a community-by-species matrix.
  
  # reformat the dataframe - wider
comm.wide.df <- sample.df %>%
    pivot_wider(names_from = taxon, values_from = count)
  
  # replace all NAs with 0
comm.wide.df[is.na(comm.wide.df)] <- 0
comm.matrix <- column_to_rownames(comm.wide.df, var = "sample")
  
  # split apart the sample name in the data frame to get the strata
  community.df.sep <- comm.wide.df %>%
    separate(sample, into = c("ablg", "plate", "replicate"), remove = FALSE)
  
  # Calculating relative abundance and creating new dataframe with relative abundance data
  relative.abund <- decostand(comm.matrix, method = "total") # standardization method = total, which probably makes the most sense because it sums over the rows (and not the columns - which would be among replicates... which is what we're curious to calculate the dissimilarity among)
  
  matrix.prop <- as.matrix(relative.abund)
  
  # jaccard similarity (presence/absence)
  jaccard.dist <- vegdist(relative.abund, method = "jaccard")
  
  # generate an easily-filtered dataframe with any comparisons that are > 0.49 dissimilar
  # Creating easy to view matrix
  jaccard_distmat2 <- as.matrix(jaccard.dist, labels = T)
  
  as.data.frame.table(jaccard_distmat2) %>%
  rename(rep1 = Var1, rep2 = Var2, Jaccard_dist = Freq) %>%
  filter(Jaccard_dist > 0.5) 
  
  # Permanova for jaccard 
  set.seed(129)
  jaccard.pool.perm <- adonis2(jaccard.dist~plate, data = community.df.sep, permutations = 999, method = "jaccard")
  
  # Bray-Curtis dissimilarity
  # Calculate distance matrix
  bray.distmat <- vegdist(relative.abund, method = "bray")
  
  # generate an easily-filtered dataframe with any comparisons that are > 0.49 dissimilar
  # Creating a matrix
  bray_distmat2 <- as.matrix(bray.distmat, labels = T)
  # reformat for easier filtering as a table
  bray_df <- as.data.frame.table(bray_distmat2) %>%
    rename(rep1 = Var1, rep2 = Var2, BrayCurtis_dist = Freq)

```
Both Jaccard and Bray-Curtis fare poorly with rare targets, I'd bet.

```{r explore-a-bit-more}
bray_reps <- bray_df %>%
  separate(rep1, into = c("ablg.1", "plate.1", "rep.1")) %>%
  separate(rep2, into = c("ablg.2", "plate.2", "rep.2") ) %>%
  filter(ablg.1 == ablg.2) %>% # only the same ablg number
  filter(plate.1 == plate.2) %>% # compare reps a and b
  filter(rep.1 != rep.2) # but no need to compare replicates with themselves

bray_reps %>%
  ggplot(aes(BrayCurtis_dist)) +
  geom_histogram()

```
The vast majority of a/b replicates are more similar than dissimilar.
258 comparisons; 24 > 0.5

```{r}
24/258
```
just under 10% (9.3%)

Quickly take a look at any that are more dissimilar:
```{r dissimilar-extraction-replicates}
bray_reps %>%
  filter(BrayCurtis_dist > 0.5) #%>%
 # unique()

```

What about plate replicates?
We would anticipate these being more similar than the extraction replicates, potentially.
```{r plate-replicates-bray}
bray_plates <- bray_df %>%
  separate(rep1, into = c("ablg.1", "plate.1", "rep.1")) %>%
  separate(rep2, into = c("ablg.2", "plate.2", "rep.2") ) %>%
  filter(ablg.1 == ablg.2) %>% # only the same ablg number
  filter(rep.1 == rep.2) %>% # compare plate replicates
  filter(plate.1 != plate.2) # but no need to compare replicates with themselves

bray_plates %>%
  ggplot(aes(BrayCurtis_dist)) +
    geom_histogram()
```

```{r}
bray_plates %>%
  filter(BrayCurtis_dist > 0.5) %>%
  select(ablg.1) %>%
  unique()
```
82/516

```{r}
82/516
```
~16% of PCR replicates were more dissimilar (from 11 samples)

What about summarizing over all replicates. We assume that detections are true detections.
```{r insert-min-read-depth}
min_read_df %>%
  ggplot(aes(taxon_reads)) +
    geom_histogram()

presence_df <- min_read_df %>%
  group_by(ablg) %>%
  mutate(presence = ifelse(taxon_reads > 10 & prop_taxon > 0.1, "1", "0")) %>%
  select(ablg, alternative_ID, source, taxon, presence) %>%
  unique() %>%
  mutate(detection_method = "DNA")
  
presence_df
```
I wonder if samples with less material (fewer total reads) have a different composition than samples with a greater number of total reads?
```{r do-lower-depth-samples-have-more-species?}
# Somewhat counterintuitive, but I'm thinking of them acting like negative controls... with lower amounts of DNA to amplify (other than predator DNA)

depth_df <- min_read_df %>%
  mutate(depth = ifelse(total_reads < 1000, "< 1,000", NA)) %>%
  mutate(depth = ifelse(total_reads > 1000 & total_reads < 5000, "< 5,000", depth)) %>%
    mutate(depth = ifelse(total_reads > 5000 & total_reads < 10000, "< 10,000", depth)) %>%
    mutate(depth = ifelse(total_reads > 10000, "> 10,000", depth)) 

depth_df %>%
  #filter(taxon_reads > 10) %>%
  ggplot(aes(x = sample, y = taxon_reads/total_reads, fill = taxon)) +
  geom_bar(stat = "identity") +
  facet_grid(cols = vars(depth), scales = "free_x", space = "free") +
    theme_bw() +
  theme(
    legend.position = "none",
    #axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    axis.text.x = element_blank()
  ) 

```
What about a species accumulation curve or similar based on the read depth?
```{r}
depth_df %>%
  group_by(sample) %>%
  add_tally(name = "n_taxa") %>%
  group_by(depth) %>%
  summarise(mean(n_taxa))
  
```
Interesting. Samples with < 5000 reads have, on average, 2 additional taxa than samples with > 10,000 reads.

What if I just look by genus?
Need to deal with some manual taxonomy stuff.

```{r}
depth_df %>%
  group_by(ablg) %>%
  mutate(presence = ifelse(taxon_reads > 10, "1", "0")) %>%
  ungroup() %>%
  select(taxon) %>%
  unique() %>%
  arrange(taxon)


```
72 taxon (species/genus/family)

Major constituents:
i.e., >10% of a given sample:
```{r}
min_read_df %>%
  filter(total_reads > 1000) %>% # samples with so few prey reads are less reliable
  filter(prop_taxon > 0.1) %>%
  ungroup() %>%
  arrange(desc(prop_taxon)) %>%
  select(taxon) %>%
  unique()

```






```{r}
tmp_df <- hardparts_df %>%
  filter(DNASample != "NULL") %>%
  mutate(detection_method = "hardparts") %>%
  rename(taxon = hardparts, alternative_ID = DNASample) %>%
  select(-HardPartsSample, -SpeciesCode, -CommonName, -Notes) %>%
  mutate(source = ifelse(str_detect(alternative_ID, "Cu23"), "scat matrix", NA)) %>%
  mutate(source = ifelse(str_detect(alternative_ID, "AN"), "fecal loop", source)) %>%
  mutate(taxon = str_replace(taxon, " sp.", "")) %>%
  mutate(taxon = str_replace(taxon, "Actinopterygii", "Actinopteri")) %>%
  mutate(taxon = str_replace(taxon, "Stenobrachius leucopsaurus", "Stenobrachius"))


# plot comparison
presence_df %>%
  bind_rows(., tmp_df) %>%
  mutate(taxon = str_replace(taxon, "Stenobrachius leucopsaurus", "Stenobrachius")) %>%
  ggplot(aes(x = alternative_ID, fill = taxon)) +
  geom_bar(stat = "count") +
  facet_grid(rows = vars(detection_method), cols = vars(source), space = "free", scales = "free") +
  theme_bw() +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  ) 
  
```

```{r}
combo_df <- presence_df %>%
    #   mutate(taxon = ifelse(taxon == "Leuroglossus", "Leuroglossus schmidti", taxon)) %>%
    #     mutate(taxon = str_replace(taxon, "Sebastes entomelas", "Sebastes")) %>%
    # mutate(taxon = str_replace(taxon, "Sebastes zacentrus", "Sebastes")) %>%
  bind_rows(., tmp_df) 
    # mutate(taxon = str_replace(taxon, "Stenobrachius leucopsaurus", "Stenobrachius")) 

 combo_df$taxon <- sub(" ", "_", combo_df$taxon)

 
 combo_df %>%
  filter(taxon != "Actinopteri") %>%
  group_by(taxon) %>%
  add_tally() %>%
  filter(source == "scat matrix") %>%
  ggplot(aes(x = reorder(taxon, -n), fill = detection_method)) +
  geom_bar(stat = "count") +
  facet_grid(rows = vars(source), space = "free", scales = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )  +
  labs(x = "taxon",
       y = "number of samples")
  
ggsave("pdf_outputs/NFSfecal_DNAhardparts.pdf", width = 8, height = 6)

# and then output a couple of plots with different taxonomic levels:
combo_df %>%
  filter(taxon != "Actinopteri") %>%
  mutate(taxon = ifelse(str_detect(taxon, "Sebastes"), "Sebastes", taxon)) %>%
  mutate(taxon = ifelse(taxon == "Hexagrammos", "Hexagrammidae", taxon)) %>%
  group_by(taxon) %>%
  add_tally() %>%
  filter(source == "scat matrix") %>%
  mutate(taxonomic_level = ifelse(str_detect(taxon, "_"), "species", "genus")) %>%
  mutate(taxonomic_level = ifelse(str_detect(taxon, "dae"), "family", taxonomic_level)) %>%
  ggplot(aes(x = reorder(taxon, -n), fill = detection_method)) +
  geom_bar(stat = "count") +
  facet_grid(cols = vars(taxonomic_level), space = "free", scales = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )  +
  labs(y = "number of samples",
       x = "",
       title = "Minimum filters: taxa with > 10 reads and >10% of reads in 1 replicate")

ggsave("pdf_outputs/NFSfecal_DNAhardparts_taxlevel.jpg", width = 10, height = 6)

```


Myctophids?
```{r}
check <- read_csv("csv_outputs/inrange_fecal_mifish_collapsed_taxonomy.csv")

all_taxa <-read_tsv("../data/NFSfecal_mifish_shorttaxlineage.txt", col_names = F)


check %>%
  left_join(., all_taxa, by = c("ASV" = "X1")) %>%
  filter(str_detect(X17, "Myctophidae")) %>%
  select(taxon, taxonomic_level) %>%
  unique()

```
Two species: Northern lanternfish and garnet lanternfish



```{r}
# separate sculpins
check %>%
  left_join(., all_taxa, by = c("ASV" = "X1")) %>%
  filter(str_detect(X17, "Cottidae")) %>%
  select(taxon, taxonomic_level) %>%
  unique()
  
```
44 ASVs that correspond to sculpins of various sorts.
8 sculpins across taxonomic levels


Don't worry about hard parts and just look at metabarcoding results:

- fecal loops vs. scats
- number of species per scat
- any additional suspected contamination?


```{r}
# combine genus/family level as appropriate
# for ease of including the legend!
min_read_df %>%
  filter(prop_taxon > 0.05) %>% # taxa that make up >1% of sequencing reads
  group_by(alternative_ID, taxon) %>%
  mutate(mean_taxon_reads = mean(taxon_reads)) %>%
  select(-taxon_reads, -sample, -plate, -replicate, -prop_taxon, -total_reads) %>%
  unique() %>%
  mutate(taxon = ifelse(str_detect(taxon, "Sebastes"), "Sebastes", taxon)) %>%
  mutate(taxon = ifelse(taxon == "Hexagrammos", "Hexagrammidae", taxon)) %>%
  mutate(taxon = ifelse(taxon == "Gadus", "Gadidae", taxon)) %>%
  ggplot(aes(x = alternative_ID, y = mean_taxon_reads, fill = taxon)) +
  geom_bar(stat = "identity") +
  facet_grid(cols = vars(source), scales = "free", space = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.title = element_blank(),
               legend.text = element_text(size = 6),
        legend.key.size = unit(0.4, 'cm'),
    title = element_text(size = 8)
  )  +
  labs(y = "Mean sequencing reads",
       x = "",
       title = "Minimum filters: taxa with >10 reads and >5% of reads")


ggsave("pdf_outputs/fecal_mifish_filtered_read_depth.jpg", width = 8, height = 5)
```
Do I see multiple species of salmon in a single scat? The might be a red flag that the 0.5% taxonomic assignment is overly confident.


```{r plot-proportions}
filtered_df <- min_read_df %>%
  mutate(taxon = ifelse(str_detect(taxon, "Sebastes"), "Sebastes", taxon)) %>%
  mutate(taxon = ifelse(taxon == "Hexagrammos", "Hexagrammidae", taxon)) %>%
  mutate(taxon = ifelse(taxon == "Gadus", "Gadidae", taxon)) %>%
  filter(prop_taxon > 0.05) %>% # taxa that make up >5% of sequencing reads in each replicate
  group_by(alternative_ID, taxon) %>%
  mutate(mean_taxon_reads = mean(taxon_reads)) %>%
  select(-taxon_reads, -sample, -plate, -replicate, -prop_taxon, -total_reads) %>%
  unique() %>%
  group_by(alternative_ID) %>%
  mutate(sample_reads = sum(mean_taxon_reads)) 

filtered_df %>%
  ggplot(aes(x = alternative_ID, y = mean_taxon_reads/sample_reads, fill = taxon)) +
  geom_bar(stat = "identity") +
  facet_grid(cols = vars(source), scales = "free", space = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.title = element_blank(),
               legend.text = element_text(size = 6),
        legend.key.size = unit(0.4, 'cm'),
    title = element_text(size = 8)
  )  +
  labs(y = "Proportion of sequencing reads",
       x = "",
       title = "Minimum filters: taxa with >10 reads and >5% of reads")

ggsave("pdf_outputs/fecal_mifish_filtered_proportions.jpg", width = 8, height = 5)

```

Some of the samples with fewer reads have more taxa... those I wouldn't believe.

Based on that filtering, how many taxa per sample?
```{r}
filtered_df %>%
  select(taxon, alternative_ID, source) %>%
  unique() %>%
  tally()

```

```{r samples-w-multiple-spp-salmon?}
filtered_df %>%
  filter(str_detect(taxon, "Oncorhynchus")) %>%
  ggplot(aes(x = alternative_ID, y = mean_taxon_reads/sample_reads, fill = taxon)) +
  geom_bar(stat = "identity") +
  facet_grid(cols = vars(source), scales = "free", space = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.title = element_blank(),
               legend.text = element_text(size = 6),
        legend.key.size = unit(0.4, 'cm'),
    title = element_text(size = 8)
  )  +
  labs(y = "Proportion of sequencing reads",
       x = "",
       title = "Salmonids with minimum filters: taxa with >10 reads and >5% of reads") +
  scale_fill_brewer(palette = "Dark2")

ggsave("pdf_outputs/fecal_mifish_filtered_salmon_reads.jpg", width = 8, height = 5)

```

```{r samples-w-multiple-gadid-species?}
filtered_df %>%
  filter(str_detect(taxon, "Gadus") | str_detect(taxon, "Gadidae")) %>%
  ggplot(aes(x = alternative_ID, y = mean_taxon_reads/sample_reads, fill = taxon)) +
  geom_bar(stat = "identity") +
  facet_grid(cols = vars(source), scales = "free", space = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.title = element_blank(),
               legend.text = element_text(size = 6),
        legend.key.size = unit(0.4, 'cm'),
    title = element_text(size = 8)
  )  +
  labs(y = "Proportion of sequencing reads",
       x = "",
       title = "Gadids with minimum filters: taxa with >10 reads and >5% of reads") +
  scale_fill_brewer(palette = "Dark2")

ggsave("pdf_outputs/fecal_mifish_filtered_gadids.jpg", width = 8, height = 5)

```


```{r dominant-taxa-by-sample}
filtered_df %>%
  filter(mean_taxon_reads/sample_reads > 0.2) %>%
  ggplot(aes(x = alternative_ID, y = mean_taxon_reads/sample_reads, fill = taxon)) +
  geom_bar(stat = "identity") +
  facet_grid(cols = vars(source), scales = "free", space = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.title = element_blank(),
               legend.text = element_text(size = 6),
        legend.key.size = unit(0.4, 'cm'),
    title = element_text(size = 8)
  )  +
  labs(y = "Proportion of sequencing reads",
       x = "",
       title = "Taxa that comprise >20% of reads with minimum filters") +
  scale_fill_brewer(palette = "Set3")

ggsave("pdf_outputs/fecal_mifish_filtered_dominant_taxa_proportion.jpg", width = 8, height = 5)

```


Using relatively stringent thresholds, how do the fecal loops compare to the scats?

```{r dominant-taxa-by-source}
filtered_df %>%
  filter(mean_taxon_reads/sample_reads > 0.2) %>%
  group_by(source, taxon) %>%
  summarise(proportion = mean(mean_taxon_reads/sample_reads)) %>%
  unique() %>%
  group_by(source) %>%
  ggplot(aes(x = source, y = proportion, fill = taxon)) +
  geom_bar(stat = "identity") +
  #facet_grid(cols = vars(source), scales = "free", space = "free") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.title = element_blank(),
               legend.text = element_text(size = 6),
        legend.key.size = unit(0.4, 'cm'),
    title = element_text(size = 8)
  )  +
  labs(y = "Proportion of sequencing reads",
       x = "",
       title = "Taxa that comprise >20% of reads with minimum filters") +
  scale_fill_brewer(palette = "Set3")
  

```

```{r %FO-data}
filtered_df %>%
  filter(source == "scat matrix") %>%
  mutate(FO_group = ifelse(str_detect(taxon, "Oncorhynchus"), "Oncorhynchus", NA)) %>%
  mutate(FO_group = ifelse(str_detect(taxon, "Stenobrachius"), "Myctophid", FO_group)) %>%
   mutate(FO_group = ifelse(str_detect(taxon, "Gadus"), "Gadus", FO_group)) %>%
   mutate(FO_group = ifelse(taxon %in% c("Leuroglossus schmidti", "Pleurogrammus monopterygius"), taxon, FO_group)) %>%
  select(alternative_ID, FO_group) %>%
  unique() %>%
  group_by(FO_group) %>%
  tally() %>%
  mutate(n_scats = 32) %>%
  mutate(`%FO` = n/n_scats)
  
# total number of scats: 32
filtered_df %>%
  filter(source == "scat matrix") %>%
  filter( str_detect(taxon, "Oncorhynchus") |
           taxon %in% c("Leuroglossus schmidti", "Pleurogrammus monopterygius") |
           str_detect(taxon, "Stenobrachius") |
           str_detect(taxon, "Gadus")) %>%
  select(alternative_ID) %>%
  unique()
```

```{r}
filtered_df %>%
  filter(source == "fecal loop")
  
  mutate(FO_group = ifelse(str_detect(taxon, "Oncorhynchus"), "Oncorhynchus", NA)) %>%
  mutate(FO_group = ifelse(str_detect(taxon, "Stenobrachius"), "Myctophid", FO_group)) %>%
   mutate(FO_group = ifelse(str_detect(taxon, "Gadus"), "Gadus", FO_group)) %>%
   mutate(FO_group = ifelse(taxon %in% c("Leuroglossus schmidti", "Pleurogrammus monopterygius"), taxon, FO_group)) %>%
  select(alternative_ID, FO_group) %>%
  unique() %>%
  group_by(FO_group) %>%
  tally() %>%
  mutate(n_loops = 11) %>%
  mutate(`%FO` = n/n_loops)

```




